File: src/main/scala/mrtjp/fengine/tiles/FETileMap.scala
package mrtjp.fengine.tiles

import mrtjp.fengine.TileCoord

import scala.collection.mutable

trait FETileMap
{
    val tileMap:mutable.Map[TileCoord, FETile] = mutable.HashMap[TileCoord, FETile]()

    def addTile(coord:TileCoord, t:FETile):Boolean =
        if (!tileMap.contains(coord)) {
            tileMap += (coord -> t)
            true
        } else false

    def getTile(coord:TileCoord):Option[FETile] = tileMap.get(coord)

    def removeTile(coord:TileCoord):Option[FETile] = tileMap.remove(coord)

    def getTileCount:Int = tileMap.size
}

File: src/main/java/mrtjp/fengine/tiles/FETileMap.java
package mrtjp.fengine.tiles;

import mrtjp.fengine.TileCoord;

import java.util.Collection;
import java.util.Optional;

/**
 * A container of tiles.
 */
public interface FETileMap {

    /**
     * Optionally returns the tile at the given coordinate, if one exists.
     *
     * @param coord The coordinate to query
     * @return An optional containing the tile if present, or Optional.empty() otherwise
     */
    Optional<FETile> getTile(TileCoord coord);

    /**
     * Returns an iterable collection containing all coord/tile pairs added to this map.
     * @return
     */
    Collection<TileMapEntry> getEntries();

    interface TileMapEntry {
        TileCoord getCoord();
        FETile getTile();
    }
}
